New Intra-Process Communication

alsora opened the pull request:
# New Intra-Process Communication 

This (big) PR implements a new Intra-Process Communication mechanism, as described in [this design document](https://github.com/ros2/design/pull/239).

The PR does not completely implement the design, i.e. it does not add support for `KEEP_ALL` and `TRANSIENT_LOCAL` quality of service.

## The general idea

The core idea is that every time a `Subscription` that uses intra-process communication is created, also an object of type `SubscriptionIntraProcess` is created. 
This new object is stored in the `IntraProcessManager` and it contains a buffer where messages can be inserted from intra-process `Publisher`s.
The object itself is `rclcpp::Waitable` so it can be added to a callback group.

When a `Publisher` that uses intra-process communication publishes a message, this is passed to the `IntraProcessManager`.
This class knows how many `Subscriptions` require this message and can deliver the message to all of them, by adding it to their buffers.

More details in the design document.

## Simplified flow

 - User calls `Node->create_subscription`.
 - The `NodeTopics` interface, if the `use_intra_process` flag is enabled, will take care of creating the `SubscriptionIntraProcess` object, adding it to the `IntraProcessManager` and to register it in the callback group.
 - The `IntraProcessManager`, through its implementation, stores utilities for getting an `IntraProcessSubscription` given its id.

From the publisher side, the flow is almost unchanged with respect to the previous one, the only difference is that intra-process messages are not stored in the mapped ring buffers (not needed anymore), but direclty delivered to the subscriptions.


## How the buffers work

From the `IntraProcessManager` point of view, a buffer is an instance of the `IntraProcessBuffer` class.
An `IntraProcessBuffer` is templated by a `MessageT` type. This can be something like `std_msgs::msg:Int`.
The `IntraProcessBuffer` provides methods for adding/consuming different variants of `MessageT` from an underlying `BufferImplementation`.
The `BufferImplementation` is the actual buffer object.

An `IntraProcessBuffer` can have a `BufferImplementation` that stores `ConstSharedPtrMessageT` or `UniquePtrMessageT`, i.e. the `BufferT` typename.

For example: an `IntraProcessBuffer` with `MessageT = std_msgs::msg:Int` can have a `BufferImplementation` with a `BufferT = std::unique_ptr<std_msgs::msg::Int>>`.
Regardless of the `BufferT` the `IntraProcessBuffer` provides method for adding shared pointer messages or unique pointer messages.

When a message is added to the `IntraProcessBuffer`, it is responsibility of the `IntraProcessManager` to make a copy if needed, in order to match the type of the underlying `BufferImplementation`.
The reason is that the `IntraProcessManager` can do a minimal number of copies since it knows all the `Subscriptions`.

By default, the type stored in the `BufferImplementation` is the one that fits better with the callback function assigned to its `Subscription`.
I.e. if the `Subscription` has a callback that does not require ownership, a `ConstSharedPtrMessageT` buffer will be created.
Note that it is possible to change that using a new `SubscriptionOption` named `IntraProcessBufferType`.
A possible use case where you want to have a type in the buffer that is different from the callback one is when the callback require ownership and the messages are big.
Having a buffer that requires ownership too means that every time a message is published, it can potentially be copied. However, if the `Subscription` is not consuming it, the copy is wasted.
Using a `BufferImplementation` that stores shared pointers allows to delay the copy to the moment when the message is extracted from the buffer.

A `BufferImplementationBase` class exists that provides an interface to the minimal methods that a Buffer has to provide (i.e. add/consume for the different data types).
The presence of the interface allows to add different `BufferImplementation`s, for example depending on the quality of service or on the intended behavior of an application.

## Future works 

 - In order to support KEEP_ALL it's sufficient to use an appropriate `BufferImplementation`.
 - In order to support TRANSIENT_LOCAL it's sufficient to store a `PublisherIntraProcess` in the `IntraProcessManager` as it's done for the subscription. This will contain only a buffer where published messages are stored. When a TRANSIENT_LOCAL subscription is added to the `IntraProcessManager`, it can retrieve old stored messages.

## Testing

The performances of this intra-process communication mechanism have been tested by us for almost a month, finding significant improvements (as discussed in the design document).

Some unit-tests are already presents, mainly to ensure that the number of copies of the messages is correct.
I was waiting for some initial feedback to add more tests.


@dgoel @raghaprasad

alsora commented on rclcpp/src/rclcpp/node_interfaces/node_topics.cpp:
```diff
@@ -140,6 +162,20 @@ NodeTopics::add_subscription(
+      ipm->get_subscription_intra_process(subscription->get_intra_process_id());
+
+    // Add to the callback group to be notified about intra-process msgs.
+    callback_group->add_waitable(subscription_intra_process);
```
This line causes applications running with Fast-RTPS to hang up after calling `rclcpp::shutdown()`.

On the other hand the following works fine
```
get_node_base_interface()->get_default_callback_group()->add_waitable(subscription_intra_process);
```

alsora replied:
Actually lately I'm not able to reproduce the error anymore. Now it works

alsora commented:
While testing this PR I found out that it gives a compilation error if you try to create a `Subscription` with a `CallbackT` different from `MessageT`.

I noticed the error in the rosbag tests.
Here `node->create_subscription` is used to create `Subscription<MessageT>` whose callback uses `rmw_serialized_msg_t`.
Note that this does not affect the actual rosbag code, but only the tests.

I fixed the errors with a PR to rosbag repo, but I would like to know in which other cases this could happen in order to address this in the rclcpp code itself.

joxoby commented on rclcpp/include/rclcpp/buffers/ring_buffer_implementation.hpp:
```diff
@@ -0,0 +1,117 @@
+
+  void enqueue(BufferT request)
+  {
+    std::unique_lock<std::mutex> lock(mutex_);
```
`std::lock_guard` is prefered

alsora replied:
Done

joxoby commented on rclcpp/include/rclcpp/buffers/intra_process_buffer.hpp:
```diff
@@ -0,0 +1,208 @@
+  RCLCPP_SMART_PTR_DEFINITIONS(IntraProcessBuffer)
+
+  using ConstMessageSharedPtr = std::shared_ptr<const MessageT>;
+  using MessageUniquePtr = std::unique_ptr<MessageT>;
```
Why not `std::unique_ptr<const MessageT>` ?

alsora replied:
Because `std::unique_ptr<MessageT>` is the one that can be used in the callbacks
https://github.com/ros2/rclcpp/blob/master/rclcpp/include/rclcpp/any_subscription_callback.hpp#L40

joxoby commented on rclcpp/include/rclcpp/buffers/ring_buffer_implementation.hpp:
```diff
@@ -0,0 +1,117 @@
+private:
+  std::vector<BufferT> ring_buffer_;
+
+  uint32_t write_;
```
Is there any particular reason to use `uint32_t` here and not `int`?

alsora replied:
The only reason is that this is an index of a vector, so I don't need the sign.

joxoby replied:
That would be an argument for `int` vs `unsigned int`. The use of `stdint.h` types should be reserved for cases in which portability is required. In this case, `uint32_t` will degrade performance in a 64-bit architecture.

https://google.github.io/styleguide/cppguide.html#Integer_Types

alsora replied:
Changed to int

joxoby commented on rclcpp/include/rclcpp/buffers/intra_process_buffer.hpp:
```diff
@@ -0,0 +1,208 @@
+  }
+
+private:
+  std::shared_ptr<BufferImplementationBase<BufferT>> buffer_;
```
Does this need to be `shared_ptr` and not `unique_ptr` ?

alsora replied:
No, I'm going to fix it.

joxoby commented on rclcpp/include/rclcpp/buffers/buffer_implementation_base.hpp:
```diff
@@ -0,0 +1,37 @@
+class BufferImplementationBase
+{
+public:
+  virtual void dequeue(BufferT & request) = 0;
```
Why just not returning `BufferT`? RVO performs better than using `std::move`

alsora replied:
I created a branch where I allow RVO https://github.com/alsora/rclcpp/commit/68ef5cba54442894a4b201a1ee3045aaa8f30553
However, the performances are slightly worst

alsora replied:
I created a branch where I allow RVO. 
Here are the differences https://github.com/alsora/rclcpp/compare/alsora/simplified_ipc...alsora:alsora/rvalue_optimization

The performances are slightly better than the current version (in 10 runs of MontBlanc benchmark application we have 178us latency and 11.3% CPU in the current against 168us latency and 10.5% CPU in the other).

Is it the suggested behavior to rely on the compiler optimization in this case?

joxoby replied:
RVO or "copy elision" became part of the C+11 standard, so it's safe to rely on it.

joxoby replied:
Correction: _copy elision_ is not guaranteed by pre C++17 ([relevant](https://blog.tartanllama.xyz/guaranteed-copy-elision/)) standards but only allowed. However, most modern compilers implement it.

alsora replied:
Merged

joxoby commented on rclcpp/include/rclcpp/create_subscription.hpp:
```diff
@@ -88,7 +89,7 @@ template<
 typename std::shared_ptr<SubscriptionT>
 create_subscription(
-  NodeT && node,
+  NodeT & node,
```
Why was `node` binded to an rvalue reference in the first place?

joxoby commented on rclcpp/include/rclcpp/create_subscription.hpp:
```diff
@@ -88,7 +89,7 @@ template<
 typename std::shared_ptr<SubscriptionT>
 create_subscription(
-  NodeT && node,
+  NodeT & node,
```
Why was this changed?

alsora replied:
Actually this and the change just below at line 104 were needed in an older version of the code.

I kept them because the corresponding `create_publisher` function does not take an r-value reference.
https://github.com/ros2/rclcpp/blob/master/rclcpp/include/rclcpp/create_publisher.hpp#L73

joxoby replied:
It's an unrelated change so I would revert it.

alsora replied:
Reverted

joxoby commented on rclcpp/include/rclcpp/create_subscription.hpp:
```diff
@@ -136,8 +137,9 @@ create_subscription(
-    use_intra_process);
-  node_topics->add_subscription(sub, options.callback_group);
+    use_intra_process,
+    options.intra_process_buffer_type);
```
Do I care about setting `options.intra_process_buffer_type` if `use_intra_process` is false?

alsora replied:
No, `intra_process_buffer_type` is used only when `use_intra_process` is true.
I also don't really like it, but the alternative I tried was to split this into 2 functions:
```
create_subscription()
if (use_intra_process)
    create_intra_process_subscription(options.intra_process_buffer_type)
```

joxoby commented on rclcpp/include/rclcpp/intra_process_manager.hpp:
```diff
@@ -230,168 +169,89 @@ class IntraProcessManager
-      typed_buffer->pop(message_sequence_number, message);
-    }
-  }
+    std::set<uint64_t> take_shared_subscription_ids;
```
`std::unordered_set` since you don't need them sorted

alsora replied:
Yes

joxoby commented on rclcpp/include/rclcpp/intra_process_manager_impl.hpp:
```diff
@@ -52,191 +51,189 @@ class IntraProcessManagerImplBase
+      subscription->use_take_shared_method();
+
+    // adds the subscription id to all the matchable publishers
+    for (auto pair : publishers_) {
```
`auto&`

alsora replied:
Done

ivanpauno commented:
@alsora Try to avoid force pushing if possible, it does the review process more difficult.

alsora commented:
Yes, sure, sorry but I have some issues with the `--signoff` option..

ivanpauno commented:
> Yes, sure, sorry but I have some issues with the --signoff option..

No problem, I still have sometimes that issue :smile:. You can sign-off them when squashing before merging.

ivanpauno commented on rclcpp/include/rclcpp/publisher.hpp:
```diff
@@ -109,21 +100,15 @@ class Publisher : public PublisherBase
     // as do_intra_process_publish takes the ownership of the message.
-    uint64_t message_seq;
     bool inter_process_publish_needed =
       get_subscription_count() > get_intra_process_subscription_count();
```
I think we have an old error here, consider:
```cpp
bool inter_process_publish_needed =
      (get_subscription_count() > get_intra_process_subscription_count()) ||
      intra_process_is_enabled_;
```

If not, it may break transient local qos policy.

ivanpauno replied:
I have send a patch for this https://github.com/ros2/rclcpp/pull/786#issue-298548085

ivanpauno commented on rclcpp/include/rclcpp/buffers/ring_buffer_implementation.hpp:
```diff
@@ -0,0 +1,115 @@
+  int write_;
+  int read_;
+  size_t length_;
+  size_t buffer_size_;
```
Maybe this names are better:
buffer_size -> capacity
length -> size
Which mimics the names used in STL containers.

Also, consider:
write_ -> write_index_
read_ -> read_index_

alsora replied:
Done.

ivanpauno commented on rclcpp/include/rclcpp/publisher.hpp:
```diff
@@ -269,15 +232,14 @@ class Publisher : public PublisherBase
-  store_intra_process_message(
+  void
+  do_intra_process_publish(
     uint64_t publisher_id,
```
The `publisher_id` is a member variable of the class, the argument in the function isn't needed.
I would just use `intra_process_publisher_id_`.

ivanpauno replied:
Actually that's my error, because you haven't changed the signature. If you change this, also do it in the other overload above.

alsora replied:
Done

ivanpauno commented on rclcpp/include/rclcpp/subscription.hpp:
```diff
@@ -159,117 +156,13 @@ class Subscription : public SubscriptionBase
-
   bool
-  matches_any_intra_process_publishers(const rmw_gid_t * sender_gid)
+  use_take_shared_method()
```
Consider:
```cpp
bool
use_take_shared_method() const
```

alsora replied:
Done

ivanpauno commented on rclcpp/include/rclcpp/subscription_intra_process.hpp:
```diff
@@ -0,0 +1,194 @@
+  }
+
+  bool
+  add_to_wait_set(rcl_wait_set_t * wait_set)
```
This method and others here, could be implemented in the base class (and live in a .cpp file).

alsora replied:
Created `subscription_intra_process_base.hpp`, `subscription_intra_process_base.cpp` and `subscription_intra_process.hpp`

alsora commented on rclcpp/include/rclcpp/buffers/intra_process_buffer.hpp:
```diff
@@ -0,0 +1,208 @@
+
+  bool use_take_shared_method() const
+  {
+    if (std::is_same<BufferT, ConstMessageSharedPtr>::value) {
```
Changed

alsora commented on rclcpp/include/rclcpp/node_interfaces/node_topics.hpp:
```diff
@@ -64,12 +64,14 @@ class NodeTopics : public NodeTopicsInterface
   void
   add_subscription(
     rclcpp::SubscriptionBase::SharedPtr subscription,
+    bool use_intra_process,
```
Changed

alsora commented on rclcpp/include/rclcpp/publisher_base.hpp:
```diff
@@ -182,18 +181,12 @@ class PublisherBase
-  RCLCPP_PUBLIC
-  mapped_ring_buffer::MappedRingBufferBase::SharedPtr
-  virtual make_mapped_ring_buffer(size_t size) const;
-
```
Done

alsora replied:
As a note: this required substituting `dynamic_pointer_cast` with `static_pointer_cast` in `create_publisher.hpp`.

wjwwood replied:
If the destructor is not virtual then it will not call the destructor of the derived class when deleting a pointer of type `PublisherBase`. Also, if the destructor is not virtual and we don't need the destructor in `Publisher<T>` (not sure if we can get away from that) then we'd want to mark the class `final`, as it's no longer safe to casually inherit from it.

Probably, we should just leave the destructor `virtual`...

alsora commented on rclcpp/src/rclcpp/publisher_base.cpp:
```diff
@@ -250,69 +242,14 @@ PublisherBase::operator==(const rmw_gid_t * gid) const
-
-    rclcpp::exceptions::throw_from_rcl_error(ret, "could not create intra process publisher");
-  }
+  intra_process_rmw_gid_ = rmw_gid_;
```
No. Removing it.

alsora commented on rclcpp/include/rclcpp/intra_process_manager_impl.hpp:
```diff
@@ -52,191 +51,189 @@ class IntraProcessManagerImplBase
-  void
-  remove_subscription(uint64_t intra_process_subscription_id)
+    SubscriptionIntraProcessBase::SharedPtr subscription;
+    rmw_qos_profile_t options;
```
`qos` is fine

alsora commented on rclcpp/include/rclcpp/intra_process_manager_impl.hpp:
```diff
@@ -52,191 +51,189 @@ class IntraProcessManagerImplBase
-    mapped_ring_buffer::MappedRingBufferBase::SharedPtr mrb,
-    size_t size)
+    PublisherBase::WeakPtr publisher;
+    rmw_qos_profile_t options;
```
Fixed

alsora commented on rclcpp/include/rclcpp/intra_process_manager_impl.hpp:
```diff
@@ -257,95 +254,64 @@ class IntraProcessManagerImpl : public IntraProcessManagerImplBase
-  std::mutex runtime_mutex_;
+    auto subscription_it = subscriptions_.find(intra_process_subscription_id);
+    if (subscription_it == subscriptions_.end()) {
+      return std::shared_ptr<SubscriptionIntraProcessBase>(nullptr);
```
Done

alsora commented on rclcpp/include/rclcpp/publisher.hpp:
```diff
@@ -82,15 +82,6 @@ class Publisher : public PublisherBase
   virtual ~Publisher()
   {}
 
-  mapped_ring_buffer::MappedRingBufferBase::SharedPtr
```
Done

alsora commented on rclcpp/include/rclcpp/subscription_factory.hpp:
```diff
@@ -117,6 +119,36 @@ create_subscription_factory(
+          std::move(buffer));
+
+      auto sub_intra_process_base_ptr =
+        std::dynamic_pointer_cast<rclcpp::SubscriptionIntraProcessBase>(sub_intra_process);
```
Done

alsora commented on rclcpp/include/rclcpp/intra_process_manager_impl.hpp:
```diff
@@ -52,191 +51,189 @@ class IntraProcessManagerImplBase
+    return true;
   }
 
+public:
```
Done

alsora commented on rclcpp/include/rclcpp/subscription_base.hpp:
```diff
@@ -181,6 +172,9 @@ class SubscriptionBase
   }
 
+  bool
+  matches_any_intra_process_publishers(const rmw_gid_t * sender_gid);
```
Done

alsora commented on rclcpp/include/rclcpp/intra_process_manager_impl.hpp:
```diff
@@ -257,95 +254,64 @@ class IntraProcessManagerImpl : public IntraProcessManagerImplBase
+    for (auto & pair : pub_to_subs_) {
+      auto publisher_id = pair.first;
+      auto ownership_subs = pair.second.take_ownership_subscriptions;
+      auto shared_subs = pair.second.take_shared_subscriptions;
```
Done

alsora commented on rclcpp/include/rclcpp/intra_process_manager.hpp:
```diff
@@ -230,168 +169,89 @@ class IntraProcessManager
-    (void)did_replace;  // Avoid unused variable warning.
+    impl_->get_subscription_ids_for_pub(
+      take_shared_subscription_ids,
+      take_owned_subscription_ids,
```
I agree with returning a reference.. What do you think should happen if the publisher_id is not present in the intra_process_manager?

ivanpauno replied:
Thinking about this again, if we return a reference to the internally stored sets, we could access them without protection (and as I commented in other place, I think that access should be protected with a mutex).
I also think we should avoid copying them.

Probably, we should refactor a little bit `IntraProcessManager` and `IntraProcessManagerImpl`.
I will think about that later. You can left this comment unaddressed for the moment (if you have any proposal that is both thread safe and avoid the copies, please go ahead).

alsora replied:
Do you think that `IntraProcessManagerImpl` is necessary ? 
What's the rationale for having the 2 separate objects in this case ?

ivanpauno replied:
I don't think it's necessary, and I think it's doing the things more complicated than they really are.
If you want, you can drop `IntraProcessManagerImpl` and implement everything in `IntraProcessManager`.

alsora commented on rclcpp/include/rclcpp/buffers/intra_process_buffer.hpp:
```diff
@@ -0,0 +1,208 @@
+  RCLCPP_SMART_PTR_DEFINITIONS(IntraProcessBuffer)
+
+  using ConstMessageSharedPtr = std::shared_ptr<const MessageT>;
+  using MessageUniquePtr = std::unique_ptr<MessageT>;
```
Done

alsora commented on rclcpp/include/rclcpp/buffers/intra_process_buffer.hpp:
```diff
@@ -0,0 +1,208 @@
+
+template<
+  typename MessageT,
+  typename BufferT = std::unique_ptr<MessageT>>
```
Done

alsora commented on rclcpp/include/rclcpp/buffers/intra_process_buffer.hpp:
```diff
@@ -0,0 +1,210 @@
+  virtual bool use_take_shared_method() const = 0;
+};
+
+template<typename MessageT>
```
Done

alsora commented on rclcpp/include/rclcpp/buffers/intra_process_buffer.hpp:
```diff
@@ -0,0 +1,210 @@
+  RCLCPP_SMART_PTR_DEFINITIONS(TypedIntraProcessBuffer)
+
+  using ConstMessageSharedPtr = std::shared_ptr<const MessageT>;
+  using MessageUniquePtr = std::unique_ptr<MessageT>;
```
Done

alsora commented on rclcpp/include/rclcpp/intra_process_manager_impl.hpp:
```diff
@@ -52,191 +51,189 @@ class IntraProcessManagerImplBase
-    if (!publisher) {
-      throw std::runtime_error("publisher has unexpectedly gone out of scope");
+    if (subscriptions_.find(id) != subscriptions_.end()) {
+      return;
```
Done

alsora commented on rclcpp/include/rclcpp/intra_process_manager_impl.hpp:
```diff
@@ -257,95 +254,64 @@ class IntraProcessManagerImpl : public IntraProcessManagerImplBase
-    if (publisher_it == publishers_.end()) {
+    auto publisher_it = pub_to_subs_.find(intra_process_publisher_id);
+    if (publisher_it == pub_to_subs_.end()) {
       // Publisher is either invalid or no longer exists.
```
Done

ivanpauno commented on rclcpp/include/rclcpp/intra_process_manager_impl.hpp:
```diff
@@ -52,191 +51,189 @@ class IntraProcessManagerImplBase
+    }
+
+    // a reliable subscription can't be connected with a best effort publisher
+    if (sub_info.options.reliability == RMW_QOS_POLICY_RELIABILITY_RELIABLE &&
```
https://community.rti.com/static/documentation/connext-dds/5.3.0/doc/manuals/connext_dds/html_files/RTI_ConnextDDS_CoreLibraries_UsersManual/Content/UsersManual/QoS_Requested_vs__Offered_Compatibility_.htm
I don't know what configuration are we using in our rmw implementations. Will check.

ivanpauno replied:
The link I send above is actually confirming that your implementation is correct.
I will double check with the others DDS vendors, but they should be doing the same.

Also, I don't know if other combinations of QoS policies are also incompatible and should be added here.

ivanpauno commented on rclcpp/include/rclcpp/intra_process_manager.hpp:
```diff
@@ -230,168 +169,89 @@ class IntraProcessManager
-    if (!typed_buffer) {
-      throw std::runtime_error("Typecast failed due to incorrect message type");
+    if (take_owned_subscription_ids.size() > 0) {
+      auto unique_msg = std::make_unique<MessageT>(*message);
```
[Here](https://github.com/ros2/rclcpp/blob/8c48dbede7de2bc7e5199f340b6bfd33ffd6a784/rclcpp/include/rclcpp/mapped_ring_buffer.hpp#L113-L116) is an example.

alsora commented on rclcpp/include/rclcpp/subscription_intra_process.hpp:
```diff
@@ -0,0 +1,194 @@
+  : any_callback_(callback), buffer_(std::move(buffer)), topic_name_(topic_name), qos_profile_(qos_profile)
+  {
+    std::shared_ptr<rclcpp::Context> context_ptr =
+      rclcpp::contexts::default_context::get_global_default_context();
```
Done

alsora commented on rclcpp/include/rclcpp/buffers/ring_buffer_implementation.hpp:
```diff
@@ -0,0 +1,115 @@
+private:
+  std::vector<BufferT> ring_buffer_;
+
+  int write_;
```
Isn't `int` suggested when I do not have to compare it with "sizes", but just increment and decrement, even if this number should not go below zero (either negative or underflowing)

ivanpauno replied:
The maximum of an `int` is less that the maximum of a `size_t` variable.
If the capacity is over `std::numeric_limits<int>::max` (which is possible, as `capacity_` is of type `size_t`), `int` indexes won't be enough to access the whole buffer.

In general, STL use `size_t` for indexes (e.g. https://en.cppreference.com/w/cpp/container/vector/at), I would do the same.

alsora replied:
Changed to `size_t`

alsora commented on rclcpp/include/rclcpp/intra_process_manager.hpp:
```diff
@@ -230,168 +169,89 @@ class IntraProcessManager
-    typename MessageT, typename Alloc = std::allocator<void>>
-  uint64_t
-  store_intra_process_message(
+  template<typename MessageT>
```
@ivanpauno Could you have a look at the `intra_process_manager.hpp` file?
I changed the `do_intra_process_publish` signature to accept a MessageAllocator (the one stored in the publisher)

ivanpauno replied:
I think it's ok, thanks!

alsora commented on rclcpp/include/rclcpp/subscription_factory.hpp:
```diff
@@ -117,6 +119,36 @@ create_subscription_factory(
+          any_subscription_callback,
+          topic_name,
+          subscription_options.qos,
+          std::move(buffer));
```
Done

alsora commented on rclcpp/include/rclcpp/subscription_factory.hpp:
```diff
@@ -55,14 +57,14 @@ struct SubscriptionFactory
+    std::function<rclcpp::SubscriptionIntraProcessBase::SharedPtr(
+        rclcpp::IntraProcessBufferType buffer_type,
+        const std::string & topic_name,
+        const rcl_subscription_options_t & subscription_options)>;
```
This is fine according to the linter

alsora commented on rclcpp/include/rclcpp/buffers/intra_process_buffer.hpp:
```diff
@@ -0,0 +1,208 @@
+  add_message(MessageUniquePtr unique_msg)
+  {
+    buffer_->enqueue(std::move(unique_msg));
+  }
```
Done the second

alsora commented on rclcpp/include/rclcpp/buffers/ring_buffer_implementation.hpp:
```diff
@@ -0,0 +1,113 @@
+
+  void dequeue(BufferT & request)
+  {
+    assert(has_data());
```
Done

alsora commented on rclcpp/src/rclcpp/node_interfaces/node_topics.cpp:
```diff
@@ -43,23 +43,25 @@ NodeTopics::create_publisher(
       throw std::invalid_argument(
               "intraprocess communication is not allowed with keep all history qos policy");
     }
-    if (publisher_options.qos.depth == 0) {
```
Actually this situation would result in throwing an error later, when the intra-process buffer is created.
Restoring this check

alsora replied:
In the current implementation the buffer is on the subscription side, so this is not needed.
However, after this PR I can add support for transient local so this will become a condition relevant only when IPC is on and durability is not volatile.

alsora commented on rclcpp/include/rclcpp/intra_process_manager_impl.hpp:
```diff
@@ -52,191 +51,189 @@ class IntraProcessManagerImplBase
 
-    return info.buffer;
+    // a publisher and a subscription with different durability can't communicate
+    if (sub_info.options.durability != pub_info.options.durability) {
```
I agree with you, but I haven't found such a function in the rmw.

I added a todo for now.
I was wondering if something like this should be implemented in the `rmw` or in the specific `rmw_implementations`...

wjwwood replied:
I think this one is a problem. It's something that was done inherently in the old intra process implementation. It should really be up to the rmw to decide this logic I think. I'll bring it up in the ROS 2 meeting.

wjwwood replied:
We decided to fix this after merging this, but we'll most likely add an rmw function for this and have a single implementation of it in `rmw` itself that we recommend all middleware implementations use, but it will still be possible for them to make their own if they wish (understanding that changing these assumptions may break tests and/or existing applications).

alsora commented:
@ivanpauno I think I addressed all your comments except for the following in `intra_process_manager_impl.hpp`:
 - pass the subscription ids as const reference instead of copying.
 - use a mutex when removing publishers

ivanpauno commented:
> * use a mutex when removing publishers

What I tried to say, is that a mutex is needed for accessing any of the maps in the intraprocess manager.
If not, I think it's not thread safe.

alsora commented:
Removed the `intra_process_manager_impl` and added mutex.

ivanpauno commented:
It is commented in the design document, that choosing the shared ptr buffer implementation is possible with a unique_ptr callback.
I think this should be changed for that to work:
https://github.com/alsora/rclcpp/blob/957d26609f8f475712c173d7d356783897735383/rclcpp/include/rclcpp/any_subscription_callback.hpp#L179-L217

Since choosing the kind of buffer is possible manually, I think L178-191, L211-214 are wrong.

Could you check please? Thanks!

ivanpauno commented:
> I also think [ros2/rosbag2#140](https://github.com/ros2/rosbag2/pull/140) shouldn't be needed.
> I will later think in how to avoid that problem.

I have tested locally, and I can build all rosbag2 packages without using https://github.com/ros2/rosbag2/pull/140, using these changes in rclcpp.

Is https://github.com/ros2/rosbag2/pull/140 needed? Or can we just close it?

alsora commented on rclcpp/src/rclcpp/intra_process_manager.cpp:
```diff
@@ -21,60 +21,133 @@ namespace intra_process_manager
-  size_t buffer_size)
+  rclcpp::PublisherBase::SharedPtr publisher)
 {
   auto id = IntraProcessManager::get_next_unique_id();
```
Done

alsora commented on rclcpp/src/rclcpp/intra_process_manager.cpp:
```diff
@@ -21,60 +21,133 @@ namespace intra_process_manager
-  rclcpp::SubscriptionBase::SharedPtr subscription)
+  SubscriptionIntraProcessBase::SharedPtr subscription)
 {
   auto id = IntraProcessManager::get_next_unique_id();
```
Done

alsora commented on rclcpp/src/rclcpp/intra_process_manager.cpp:
```diff
@@ -21,60 +21,133 @@ namespace intra_process_manager
 IntraProcessManager::matches_any_publishers(const rmw_gid_t * id) const
 {
-  return impl_->matches_any_publishers(id);
+  for (auto & publisher_pair : publishers_) {
```
Done

alsora commented on rclcpp/src/rclcpp/intra_process_manager.cpp:
```diff
@@ -21,60 +21,133 @@ namespace intra_process_manager
 IntraProcessManager::get_subscription_count(uint64_t intra_process_publisher_id) const
 {
-  return impl_->get_subscription_count(intra_process_publisher_id);
+  auto publisher_it = pub_to_subs_.find(intra_process_publisher_id);
```
Done

alsora commented on rclcpp/include/rclcpp/create_intra_process_buffer.hpp:
```diff
@@ -0,0 +1,93 @@
+
+template<
+  typename MessageT,
+  typename Alloc>
```
Done

alsora commented on rclcpp/include/rclcpp/intra_process_manager.hpp:
```diff
@@ -230,168 +169,127 @@ class IntraProcessManager
+    } else if (sub_ids.take_ownership_subscriptions.size() > 0 &&
+      sub_ids.take_shared_subscriptions.size() > 1)
+    {
+      std::shared_ptr<MessageT> shared_msg = std::make_shared<MessageT>(*message);
```
Done

alsora commented on rclcpp/include/rclcpp/buffers/intra_process_buffer.hpp:
```diff
@@ -0,0 +1,197 @@
+  consume_unique_impl()
+  {
+    MessageSharedPtr buffer_msg = buffer_->dequeue();
+    return std::make_unique<MessageT>(*buffer_msg);
```
Done

alsora commented on rclcpp/test/test_publisher.cpp:
```diff
@@ -198,10 +198,6 @@ static std::vector<TestParameters> invalid_qos_profiles()
     TestParameters(
       rclcpp::QoS(rclcpp::KeepLast(10)).transient_local(),
       "transient_local_qos"));
-  parameters.push_back(
```
Done

alsora commented:
@ivanpauno I addressed your comments.

> It is commented in the design document, that choosing the shared ptr buffer implementation is possible with a unique_ptr callback.

This is possible. It's handled by the `execute` method of the `subscription_intra_process`. This object will request the correct type from the `intra_process_buffer` depending on the callback ownership.



> Is [ros2/rosbag2#140](https://github.com/ros2/rosbag2/pull/140) needed? Or can we just close it?

It's still needed. You can build fine, but you can't run the tests.
That PR addresses 2 things: first  it avoids creating an intra-process subscription with a serialized_t callback, since it is not supported, but also it avoids calling some API that do not exists anymore (mainly related to mapped ring buffer)

alsora commented:
P.S.
I have the linter failing in a couple places and I have no idea about how to fix it.
```
ros2/rclcpp/rclcpp/include/rclcpp/intra_process_manager.hpp:272:  If an else has a brace on one side, it should have it on both  [readability/braces] [5]
```
This happens also at line 289 of the same file...  If I keep it on the same line then it complains that it's more than 100 characters, but it complains also if I do
```
}
else if ...
{
```

Similarly, in `ros2/rclcpp/rclcpp/src/rclcpp/node_interfaces/node_topics.cpp:122`.. It requires a lot of indentation for the text of throw, but this ends up exceeding 100 characters..

ivanpauno commented:
> @ivanpauno I addressed your comments.
> 
> > It is commented in the design document, that choosing the shared ptr buffer implementation is possible with a unique_ptr callback.
> 
> This is possible. It's handled by the `execute` method of the `subscription_intra_process`. This object will request the correct type from the `intra_process_buffer` depending on the callback ownership.
> See [here](https://github.com/alsora/rclcpp/blob/alsora/simplified_ipc/rclcpp/include/rclcpp/subscription_intra_process.hpp#L90)
> The `subscription_intra_process` can "consume" the data in the buffer without knowing how it's stored, since it has to call type-specific methods (`consume_unique`, `consume_shared`). The alternative is to allow only to consume for the stored type and then let the `any_callback` handle it.
> Since now the intra-process buffer is created inside the `subscription_intra_process`, also the second option can be done.

Yes, you're right. Thanks for the clarification.

> > Is [ros2/rosbag2#140](https://github.com/ros2/rosbag2/pull/140) needed? Or can we just close it?
> 
> It's still needed. You can build fine, but you can't run the tests.
> That PR addresses 2 things: first it avoids creating an intra-process subscription with a serialized_t callback, since it is not supported, but also it avoids calling some API that do not exists anymore (mainly related to mapped ring buffer)

I think the changes for avoiding the old api changes are needed.
But I think that creating a subscription with a serialized_t callback should be possible by doing `node->create_subscription(...`. I will try later to dig more into the problem.

alsora commented:
Is it correct to assume that if the callback type is `serialized_t` then Intra-Process Communication is also disabled?

ivanpauno commented:
> Is it correct to assume that if the callback type is `serialized_t` then Intra-Process Communication is also disabled?

Considering that intraprocess is configurable at publisher/subscription level, I would raise an error when the callback type is `serialized_t` and intraprocess is enabled.
In https://github.com/ros2/rosbag2/pull/140, I think you can explicitly disable intraprocess using [SubscriptionOptions](https://github.com/alsora/rclcpp/blob/26f2c233c7b712674df09b4bd12169bf381a776f/rclcpp/include/rclcpp/subscription_options.hpp#L33-L45), and then the usual `node->create_subscription` call.

alsora commented on rclcpp/include/rclcpp/intra_process_manager.hpp:
```diff
@@ -230,168 +170,138 @@ class IntraProcessManager
+
+      this->template add_shared_msg_to_buffers<MessageT>(msg, sub_ids.take_shared_subscriptions);
+    } else if (!sub_ids.take_ownership_subscriptions.empty() &&
+      sub_ids.take_shared_subscriptions.size() <= 1)
```
Done

alsora replied:
@wjwwood  your suggestion fails with the following errors:

```
40: @@ -274,2 +274 @@
40: -      sub_ids.take_shared_subscriptions.size() <= 1
40: -    )
40: +      sub_ids.take_shared_subscriptions.size() <= 1)
```

```
/root/rclcpp/rclcpp/include/rclcpp/intra_process_manager.hpp:272:  If an else has a brace on one side, it should have it on both  [readability/braces] [5]
```

alsora commented on rclcpp/src/rclcpp/node_interfaces/node_topics.cpp:
```diff
@@ -115,6 +115,12 @@ NodeTopics::create_subscription(
               "intraprocess communication is not allowed with keep all history qos policy");
     }
+    if (subscription_options.qos.history == RMW_QOS_POLICY_HISTORY_KEEP_LAST
+        && subscription_options.qos.depth == 0)
```
Done

alsora commented on rclcpp/include/rclcpp/buffers/intra_process_buffer.hpp:
```diff
@@ -0,0 +1,197 @@
+template<
+  typename MessageT,
+  typename Alloc = std::allocator<void>,
+  typename BufferT = std::unique_ptr<MessageT>>
```
Added deleter template argument

alsora commented:
@ivanpauno I tried different things for the rosbag problem and I committed the only one that was working, even if I don't really like it.

For the rosbag repository I have a new PR with only the API changes

ivanpauno commented:
> @ivanpauno I tried different things for the rosbag problem and I committed the only one that was working, even if I don't really like it.
> 
> For the rosbag repository I have a new PR with only the API changes

Could you copy the PR link? I can't find it in `rosbag2`.

alsora commented:
My fault, I tagged you there now.

ivanpauno commented on rclcpp/include/rclcpp/subscription_factory.hpp:
```diff
@@ -117,6 +119,32 @@ create_subscription_factory(
+      }
+
+      auto sub_intra_process =
+        std::make_shared<SubscriptionIntraProcess<MessageT, Alloc, std::default_delete<MessageT>, CallbackMessageT>>(
```
Classes taking an allocator type as template parameter don't need a deleter too (the deleter can be deduced from the allocator type, e.g.: https://github.com/ros2/rclcpp/blob/9be3e08cd4d04193fb76cb6dd8f5db11d7199e2f/rclcpp/include/rclcpp/mapped_ring_buffer.hpp#L64-L66
A deleter template parameter is needed in functions with an `unique_ptr` argument (or return value) that takes it (it could be sth different of a `unique_ptr` too, any class needing a deleter), or in a class that needs to know about the deleter and not the allocator type (which I think doesn't happen here).
You usually don't need to pass the deleter template parameter explicitly (because it's automatically deduced). This case is probably an exception of that, so pass a deleter deduced from the allocator type (`create_subscription_factory` is a template taking an allocator type) and not `std::default_delete`.

alsora replied:
Updated the template argument.. I'm not very familiar with C++ allocator concepts, sorry that it required so many changes.

ivanpauno replied:
Allocator usage looks correct now.

alsora commented:
I found the method in Fast-RTPS for matching the QOS
https://github.com/eProsima/Fast-RTPS/blob/master/src/cpp/rtps/builtin/discovery/endpoint/EDP.cpp#L457

ivanpauno commented:
Running CI (up to rclcpp, test_rclcpp, tlsf_cpp, demo_nodes_cpp, intra_process_demo, rosbag2), to catch failures before new review:

* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=7725)](http://ci.ros2.org/job/ci_linux/7725/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=3829)](http://ci.ros2.org/job/ci_linux-aarch64/3829/)
* macOS [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_osx&build=6326)](http://ci.ros2.org/job/ci_osx/6326/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=7579)](http://ci.ros2.org/job/ci_windows/7579/)

ivanpauno commented:
> The RMW would need a function that takes as argument topic name and QoS object.
> Then each implementation can convert these in an appropriate structure and use the method already defined in the DDS library

Reusing a DDS library function is the best option. If a free function isn't available, doing our implementation and a test checking that it behaves exactly in the same way as in the DDS vendor would be fine.

alsora commented:
My branch was not up-to-date with the rclcpp master.. Now it builds fine

ivanpauno commented:
Again:
* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=7728)](http://ci.ros2.org/job/ci_linux/7728/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=3832)](http://ci.ros2.org/job/ci_linux-aarch64/3832/)
* macOS [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_osx&build=6328)](http://ci.ros2.org/job/ci_osx/6328/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=7587)](http://ci.ros2.org/job/ci_windows/7587/)

ivanpauno commented:
@alsora Could you check `tlsf_cpp` build errors?

alsora commented:
@ivanpauno I had to make some small fixes:
 - In a couple of places the custom deleter was not passed as template argument
 - The `rmw_message_info_t` object associated with intra-process messages was not setting the `from_intra_process` flag to true.

Moreover there are other repositories that were using the `IntraProcessManagerIml` (that now does not exist anymore).

In summary, this PR requires changes to the following repositories:
 - demos https://github.com/ros2/demos/pull/382
 - realtime_support https://github.com/ros2/realtime_support/pull/80
 - rosbag2 https://github.com/ros2/rosbag2/pull/143

ivanpauno commented:
> * The `rmw_message_info_t` object associated with intra-process messages was not setting the `from_intra_process` flag to true.

Thanks for the change.

> Moreover there are other repositories that were using the `IntraProcessManagerIml` (that now does not exist anymore).

Now I remember why we had an `IntraProcessManagerImpl`: Passing an allocator to it.
IMO, it's not really important as all the allocations done when publishing (and handling a subscription) are done with an allocator. All the allocations that are done without an allocator are solved when creating a publisher/subscriber (which I think it's fine).
For now, wait for the second reviewer opinion.

----

CI:
* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=7738)](http://ci.ros2.org/job/ci_linux/7738/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=3841)](http://ci.ros2.org/job/ci_linux-aarch64/3841/)
* macOS [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_osx&build=6833)](http://ci.ros2.org/job/ci_osx/6328/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=7597)](http://ci.ros2.org/job/ci_windows/7597/)

alsora commented:
Don't know why `test_intra_process_manager` failed now while it was always building nicely.
However I updated few things in that test and now it should not have any problems

ivanpauno commented:
> Don't know why `test_intra_process_manager` failed now while it was always building nicely.
> However I updated few things in that test and now it should not have any problems

mmm I don't know exactly, but the changes seem to be fine. Again:

* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=7741)](http://ci.ros2.org/job/ci_linux/7741/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=3844)](http://ci.ros2.org/job/ci_linux-aarch64/3844/)
* macOS [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_osx&build=6334)](http://ci.ros2.org/job/ci_osx/6334/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=7600)](http://ci.ros2.org/job/ci_windows/7600/)

alsora commented:
Updated rosbag PR with the missing semicolumn

ivanpauno commented:
* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=7743)](http://ci.ros2.org/job/ci_linux/7743/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=3846)](http://ci.ros2.org/job/ci_linux-aarch64/3846/)
* macOS [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_osx&build=6336)](http://ci.ros2.org/job/ci_osx/6336/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=7600)](http://ci.ros2.org/job/ci_windows/7600/)

alsora commented:
This package is failing in the macOS and WIndows builds.
In the macOS the problem looks like I can't use `RCLCPP_SMART_PTR_DEFINITIONS` on a virtual class.
In Windows the test `test_subscription_publisher_count_api` has unresolved external symbols `rclcpp::SubscriptionBase::matches_any_intra_process_publishers` and `rclcpp::SubscriptionIntraProcessBase::add_to_wait_set`

ivanpauno commented:
> This package is failing in the macOS and WIndows builds.
> In the macOS the problem looks like I can't use `RCLCPP_SMART_PTR_DEFINITIONS` on a virtual class.

Use `RCLCPP_SMART_PTR_ALIASES_ONLY` instead:

https://github.com/ros2/rclcpp/blob/9be3e08cd4d04193fb76cb6dd8f5db11d7199e2f/rclcpp/include/rclcpp/macros.hpp#L56-L69

> In Windows the test `test_subscription_publisher_count_api` has unresolved external symbols `rclcpp::SubscriptionBase::matches_any_intra_process_publishers` and `rclcpp::SubscriptionIntraProcessBase::add_to_wait_set`

Double check if you used `RCLCPP_PUBLIC` macro.

alsora commented:
Updated using these macros

ivanpauno commented on rclcpp/include/rclcpp/intra_process_manager.hpp:
```diff
@@ -400,18 +310,122 @@ class IntraProcessManager
-  std::mutex take_mutex_;
+  void insert_sub_id_for_pub(uint64_t sub_id, uint64_t pub_id, bool use_take_shared_method);
+
+  bool can_communicate(PublisherInfo pub_info, SubscriptionInfo sub_info) const;
```
Add `RCLCPP_PUBLIC`

ivanpauno commented on rclcpp/include/rclcpp/intra_process_manager.hpp:
```diff
@@ -400,18 +310,122 @@ class IntraProcessManager
 
-  IntraProcessManagerImplBase::SharedPtr impl_;
-  std::mutex take_mutex_;
+  void insert_sub_id_for_pub(uint64_t sub_id, uint64_t pub_id, bool use_take_shared_method);
```
Add `RCLCPP_PUBLIC`

ivanpauno commented on rclcpp/include/rclcpp/subscription_base.hpp:
```diff
@@ -163,8 +152,10 @@ class SubscriptionBase
-    const rcl_subscription_options_t & intra_process_options);
+    IntraProcessManagerWeakPtr weak_ipm);
+
+  uint64_t
```
Add `RCLCPP_PUBLIC`

ivanpauno commented on rclcpp/include/rclcpp/subscription_intra_process_base.hpp:
```diff
@@ -0,0 +1,78 @@
+public:
+  RCLCPP_SMART_PTR_ALIASES_ONLY(SubscriptionIntraProcessBase)
+
+  SubscriptionIntraProcessBase(const std::string & topic_name, rmw_qos_profile_t qos_profile)
```
Add `RCLCPP_PUBLIC`

ivanpauno commented on rclcpp/include/rclcpp/subscription_intra_process_base.hpp:
```diff
@@ -0,0 +1,78 @@
+  : topic_name_(topic_name), qos_profile_(qos_profile)
+  {}
+
+  size_t
```
Add `RCLCPP_PUBLIC`

ivanpauno commented on rclcpp/include/rclcpp/subscription_intra_process_base.hpp:
```diff
@@ -0,0 +1,78 @@
+  SubscriptionIntraProcessBase(const std::string & topic_name, rmw_qos_profile_t qos_profile)
+  : topic_name_(topic_name), qos_profile_(qos_profile)
+  {}
+
```
Add virtual destructor

ivanpauno commented on rclcpp/include/rclcpp/subscription_intra_process_base.hpp:
```diff
@@ -0,0 +1,78 @@
+  virtual bool
+  use_take_shared_method() const = 0;
+
+  const char *
```
Add `RCLCPP_PUBLIC`

ivanpauno commented on rclcpp/include/rclcpp/subscription_intra_process_base.hpp:
```diff
@@ -0,0 +1,78 @@
+  const char *
+  get_topic_name() const;
+
+  rmw_qos_profile_t
```
Add `RCLCPP_PUBLIC`

alsora commented on rclcpp/include/rclcpp/macros.hpp:
```diff
@@ -66,6 +66,7 @@
 #define RCLCPP_SMART_PTR_ALIASES_ONLY(...) \
   __RCLCPP_SHARED_PTR_ALIAS(__VA_ARGS__) \
   __RCLCPP_WEAK_PTR_ALIAS(__VA_ARGS__) \
+  __RCLCPP_UNIQUE_PTR_ALIAS(__VA_ARGS__) \
```
The `RCLCPP_SMART_PTR_ALIASES_ONLY` was missing the alias for unique pointers

ivanpauno commented:
* macOS [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_osx&build=6349)](http://ci.ros2.org/job/ci_osx/6349/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=7613)](http://ci.ros2.org/job/ci_windows/7613/)

alsora commented:
Windows has the following tests failing:
```
17:06:10 build\launch\pytest.xml: 154 tests, 0 errors, 1 failure, 0 skipped
17:06:10 build\ros2component\pytest.xml: 6 tests, 0 errors, 1 failure, 0 skipped
17:06:10 build\sros2\pytest.xml: 26 tests, 0 errors, 1 failure, 0 skipped
17:06:10 build\tf2_ros\test_results\tf2_ros\tf2_ros_test_message_filter.gtest.xml: 1 test, 0 errors, 1 failure, 0 skipped
17:06:10 build\tf2_ros\testing\20190801-1547\Test.xml: 1 test, 0 errors, 1 failure, 0 skipped
```

In macOS I forgot to use the `RCLCPP_SMART_PTR_ALIASES_ONLY` macro in one of the tests.. The last commit fixes this.

ivanpauno commented:
> Windows has the following tests failing:

No problem, they are all known failures.

> In macOS I forgot to use the `RCLCPP_SMART_PTR_ALIASES_ONLY` macro in one of the tests.. The last commit fixes this.

* macOS [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_osx&build=6351)](http://ci.ros2.org/job/ci_osx/6351/)

alsora commented:
Any updates on this?

wjwwood commented:
Sorry @alsora, it's waiting on me to review it. I've been on vacation and stuff, but I'm going to try and look at it very soon.

alsora commented:
@wjwwood I fixed some conflicts due to recent changes in `rclcpp`. 
Do you had time for looking at this PR?

alsora commented on rclcpp/include/rclcpp/publisher.hpp:
```diff
@@ -109,21 +98,15 @@ class Publisher : public PublisherBase
-      message_seq =
-        store_intra_process_message(intra_process_publisher_id_, shared_msg);
+      std::shared_ptr<MessageT> shared_msg = std::move(msg);
+      this->do_intra_process_publish(shared_msg);
```
This will require to have some of the logic of intra-process communication also in the `publisher.hpp` file instead of only in intra-process-manager.
Is it worth?

wjwwood replied:
Sorry, forgot to come back to this one.

Yes, I think it's worth it, but maybe we could reorganize the API to avoid more coupling.

Right now, I believe due to this, the tutorials behind the `intra_process_demos` package no longer work correctly. As they expect the same pointer to be used iteration to iteration.

wjwwood replied:
I changed the behavior in 3013e02, please have a look if you have time. I'll  probably push ahead with that change for now, but I can fix it up later if you think it needs it.

With that change the `cyclic_pipeline` demo works again.

alsora commented:
@wjwwood I'm going to fix the conflicts and update the APIs.
I think that the only relevant change consists in changing how the intra-process subscription is created.
At the moment this is done through the subscription factory in `node_topics.cpp`.
It has to be moved to within the subscription itself in the `post_init_setup`

wjwwood commented:
Please hold off, I'm already do this rebase.

alsora commented:
Ok great, let me know if you need anything

wjwwood commented:
I rebased and in the process squashed your work so far @alsora. I did the squash just because resolving the conflicts had to  be done on each commit that touched something that changed, which was basically all of them. By squashing first I only had to resolve conflicts once.

I placed your original branch here:

https://github.com/ros2/rclcpp/tree/alsora/simplified_ipc-original

I'm still working on follow up commits to finish "fixing" all the things broken by the upstream changes. I'll get you to review them when I've pushed them.

I also have a few other things I wanted to change or at least talk to you about changing. I'll follow up on those separately.

wjwwood commented:
So one major thing I've found, since you guys decided to remove the impl class for the intra-process manager..., is that it is no longer clear to me how this demo should work/change:

https://github.com/ros2/demos/blob/2dcbf6488b918387c3f5dab997ab08c84dbf42bb/demo_nodes_cpp/src/topics/allocator_tutorial.cpp#L154-L164

@alsora if you have any thoughts on this I'd appreciate it. I'm looking into a work around for now.

----

There's a list of other things that I think need to change, but after talking with people at OR I've decided to push a lot of the new code in this pr into an `experimental` namespace so that we can get the feature in, but later I will refactor them and then move them into the public API.

wjwwood commented:
This doc needs to be updated:

https://index.ros.org/doc/ros2/Tutorials/Allocator-Template-Tutorial/#passing-an-allocator-to-the-intra-process-pipeline

wjwwood commented on rclcpp/include/rclcpp/intra_process_manager_impl.hpp:
```diff
@@ -257,95 +254,64 @@ class IntraProcessManagerImpl : public IntraProcessManagerImplBase
-
-  PublisherMap publishers_;
-
-  std::mutex runtime_mutex_;
```
The impl was removed, but does this comment still stand @ivanpauno? If not, can you resolve this thread?

wjwwood commented on rclcpp/include/rclcpp/intra_process_manager_impl.hpp:
```diff
@@ -257,95 +254,64 @@ class IntraProcessManagerImpl : public IntraProcessManagerImplBase
-    return ret;
+
+    take_shared_ids = publisher_it->second.take_shared_subscriptions;
+    take_owned_ids = publisher_it->second.take_ownership_subscriptions;
```
Is this resolved?

ivanpauno commented:
> So one major thing I've found, since you guys decided to remove the impl class for the intra-process manager..., is that it is no longer clear to me how this demo should work/change:
> 
> https://github.com/ros2/demos/blob/2dcbf6488b918387c3f5dab997ab08c84dbf42bb/demo_nodes_cpp/src/topics/allocator_tutorial.cpp#L154-L164

Yeah, I realized about that after we already had deleted it.
We should restore the implementation class.

alsora commented:
What about an API on the publisher side: 
```
set_intra_process_allocator(std::shared_ptr<MessageAlloc> allocator)
```
Alternatively this intra-process allocator can be an argument of the constructor.

Then when calling `do_intra_process_publish` you can pass it to the intra-process manager

wjwwood commented:
That doesn't account for the unordered maps in the intra process manager, which should be passed the allocator when declared (as a template argument) and initialized (as a normal argument).

> Alternatively this intra-process allocator can be an argument of the constructor.

That would work, but it has to be structured as I had it before, because you cannot make the allocator a template argument of the intra process manager, because then anyone wishing to retrieve it from the context must know which allocator it was created with, otherwise a second intra process manager would be created. That's why there's an impl class which has a base type (which is not templated) and a derived class which is templated on the allocator. The constructor of the intra process manager takes the impl base class as an argument, and therefore the intra process manager class does not need to be templated.

wjwwood commented:
I rebased again after several other rclcpp pr's caused some issues with merging.

wjwwood commented:
@Karsten1987 1c9d534 FYI, nothing for you to do, but I just wanted to let you know in case you had input.

Karsten1987 commented on rclcpp/src/rclcpp/intra_process_manager.cpp:
```diff
@@ -12,69 +12,153 @@
-  return impl_->matches_any_publishers(id);
+  std::shared_lock<std::shared_timed_mutex> lock(mutex_);
+
+  for (auto & publisher_pair : publishers_) {
```
could this loop be potentially replaced by a call to `std::find_if`?

wjwwood replied:
Perhaps? I wasn't going to ask them to change that or change it myself. I think I'll pass on this for now in the interest of getting it merged. I'll add it to my fixup list however.

wjwwood commented on rclcpp/include/rclcpp/experimental/buffers/intra_process_buffer.hpp:
```diff
@@ -0,0 +1,241 @@
+public:
+  RCLCPP_SMART_PTR_DEFINITIONS(TypedIntraProcessBuffer)
+
+  using MessageAllocTraits = allocator::AllocRebind<MessageT, Alloc>;
```
No reason, I only changed the one in LoanedMessage because it wasn't compiling.

wjwwood commented on rclcpp/include/rclcpp/experimental/buffers/intra_process_buffer.hpp:
```diff
@@ -0,0 +1,241 @@
+  explicit
+  TypedIntraProcessBuffer(
+    std::unique_ptr<BufferImplementationBase<BufferT>> buffer_impl,
+    std::shared_ptr<Alloc> allocator = nullptr)
```
I think out of scope. I already plan to remove the shared pointer to allocators, but in the mean time I would prefer to keep it consistent.

wjwwood commented on rclcpp/include/rclcpp/experimental/buffers/intra_process_buffer.hpp:
```diff
@@ -0,0 +1,241 @@
+    if (!allocator) {
+      message_allocator_ = std::make_shared<MessageAlloc>();
+    } else {
+      message_allocator_ = std::make_shared<MessageAlloc>(*allocator.get());
```
Usually this is required if `allocator` has a different type from `MessageAlloc`.

wjwwood commented on rclcpp/include/rclcpp/experimental/buffers/ring_buffer_implementation.hpp:
```diff
@@ -0,0 +1,124 @@
+#include "rclcpp/macros.hpp"
+#include "rclcpp/visibility_control.hpp"
+
+#include "iostream"
```
Looks like it was left over from debugging or something, I removed it.

wjwwood commented on rclcpp/include/rclcpp/experimental/buffers/ring_buffer_implementation.hpp:
```diff
@@ -0,0 +1,124 @@
+
+  BufferT dequeue()
+  {
+    if (!has_data()) {
```
I'll move the lock guard, but I think this class may have other thread-safety issues, like comparing size and capacity in `is_full`. So I'll add it to my list to revisit.

wjwwood commented:
@Karsten1987 or @jacobperron if either of you have time to review my latest changes I'd appreciate it.

CI:

* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=8482)](http://ci.ros2.org/job/ci_linux/8482/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=4370)](http://ci.ros2.org/job/ci_linux-aarch64/4370/)
* macOS [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_osx&build=6894)](http://ci.ros2.org/job/ci_osx/6894/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=8383)](http://ci.ros2.org/job/ci_windows/8383/)

wjwwood commented:
CI after a fix:

* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=8484)](http://ci.ros2.org/job/ci_linux/8484/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=4372)](http://ci.ros2.org/job/ci_linux-aarch64/4372/)
* macOS [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_osx&build=6896)](http://ci.ros2.org/job/ci_osx/6896/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=8388)](http://ci.ros2.org/job/ci_windows/8388/)

wjwwood commented:
CI after another other fix:

* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=8485)](http://ci.ros2.org/job/ci_linux/8485/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=4373)](http://ci.ros2.org/job/ci_linux-aarch64/4373/)
* macOS [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_osx&build=6897)](http://ci.ros2.org/job/ci_osx/6897/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=8388)](http://ci.ros2.org/job/ci_windows/8388/)

wjwwood commented:
More CI:

* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=8486)](http://ci.ros2.org/job/ci_linux/8486/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=4374)](http://ci.ros2.org/job/ci_linux-aarch64/4374/)
* macOS [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_osx&build=6898)](http://ci.ros2.org/job/ci_osx/6898/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=8389)](http://ci.ros2.org/job/ci_windows/8389/)

wjwwood commented:
CI got cut short and I fixed an issue with custom allocators:

* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=8488)](http://ci.ros2.org/job/ci_linux/8488/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=4376)](http://ci.ros2.org/job/ci_linux-aarch64/4376/)
* macOS [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_osx&build=6902)](http://ci.ros2.org/job/ci_osx/6902/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=8390)](http://ci.ros2.org/job/ci_windows/8390/)

wjwwood commented:
I linked to the restarted macOS job. I think the ci is clean and so with a small follow up review I think this is good to merge.

wjwwood commented:
Thanks for wrapping it up for me @clalancette, and thanks @alsora for working on it in the first place.

I have several follow up items to do here, but not necessarily for Eloquent. I'll follow up with new issues as soon as I have time.
